src/audio/crossover/crossover.c

```diff
@@ -617,7 +619,7 @@ static int crossover_copy(struct comp_dev *dev)
 	}
 
 	/* Check if source is active */
-	if (source_c->source->state != dev->state) {
+	if (source_c->source && source_c->source->state != dev->state) {
```

Comment:
this one actually seems wrong. It means, that `if (source_c->source == NULL)` we proceed below and try to copy - with a NULL source buffer. Shouldn't this be `if (!source_c->source || source_c->source->state != dev->state)`?

Reply:
is the case source_c->source == NULL an error case? 

Maybe we should just skip the processing in that case, instead of returning -EINVAL as suggested.

It's really confusing that we've introduced a case where the pointers are invalid and don't have a clear policy on how to deal with such cases... I feel that if we keep on arguing on each and every change it'll take eons to get it right.

@lyakh @lgirdwood this really calls for a complete review of this case and possibly the introduction of macros/helpers to deal with this invalid pointer case, otherwise everyone in the team will screw up at some point even with the best intentions.

Reply:
We did originally have a ```must check``` on the coherent release() and that could come back. I would also make the return value on release a poison value (NOT NULL for Intel) as this would make any deref obvious from the logs.

Reply:
and a poison value would also force the usage of helper macros for convenience.

Furthermore a lot of this M:N routing/walking will go further up the stack as the module API is used by more components. i.e. we will eventually have this sort of logic in one place.  @ranj063 for additional comments.